### PR TITLE
DAML LF: expose generic equality in LF 

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances the state of #3766.

With this PR, daml-lf exposes generic equality. 

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala [hunk 1]
```diff
@@ -1285,7 +1285,7 @@ class SBuiltinTest extends FreeSpec with Matchers with TableDrivenPropertyChecks
 
   }
 
-  "EQUAL_CONTRACT_ID" - {
+  "EQUAL @(TypeRep)" - {
```
Comment (associahedron) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala]:
```suggestion
  "EQUAL @TypeRep" - {
```

File: compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs [hunk 2]
```diff
@@ -783,4 +786,4 @@ decodeNM
   => (NM.Name b -> Error) -> (a -> Decode b) -> V.Vector a -> Decode (NM.NameMap b)
 decodeNM mkDuplicateError decode1 xs = do
   ys <- traverse decode1 (V.toList xs)
-  either (throwError . mkDuplicateError) pure $ NM.fromListEither ys
+  either (throwError . mkDuplicateError) pure $ NM.fromList
```
Comment (associahedron) [inline compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs]: This looks like an accidental change?

Comment (remyhaemmerle-da) [reply compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs]: Typo. The CI catches it also.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala]: fixed

**Pull request merged**